File: eng/go/canon.go
package eng

import (
	"strconv"
	"strings"
)

// canonString renders a String payload as parseable AQL source — the
// round-trip half of the canon contract. The plain single-quoted form
// is kept verbatim for ordinary content (the form every spec row and
// doc example pins); content containing a single quote switches to
// double quotes; content with both quote kinds, backslashes, or
// control characters falls back to single quotes with backslash
// escapes. Whatever is emitted re-parses to the same string —
// previously `it's` rendered as 'it's', which re-parsed wrongly.
func canonString(s string) string {
	hasSingle := strings.ContainsRune(s, '\'')
	hasEscape := strings.ContainsAny(s, "\\\n\t\r")
	switch {
	case !hasSingle && !hasEscape:
		return "'" + s + "'"
	case !strings.ContainsRune(s, '"') && !hasEscape:
		return `"` + s + `"`
	default:
		var b strings.Builder
		b.WriteByte('\'')
		for _, r := range s {
			switch r {
			case '\\':
				b.WriteString(`\\`)
			case '\'':
				b.WriteString(`\'`)
			case '\n':
				b.WriteString(`\n`)
			case '\t':
				b.WriteString(`\t`)
			case '\r':
				b.WriteString(`\r`)
			default:
				b.WriteRune(r)
			}
		}
		b.WriteByte('\'')
		return b.String()
	}
}

// Canon renders a stack of values as canonical AQL source — a string
// that, when parsed and evaluated, reproduces the input stack. Where it
// diverges from Value.String:
//
//   - atoms render as `name/q` (bare `name` would parse as a word
//     lookup, not as an atom value; the /q suffix is the preferred
//     short form over `(quote name)`)
//   - quoted lists render as `(quote [...])` so the Quoted flag survives
//     a round-trip (the /q suffix is only defined for words)
//
// Lists and maps are space-separated in both Canon and Value.String
// (commas are optional in AQL source and the default render omits
// them); the atom and quoted-list rules above are what keep Canon
// distinct from Value.String.
//
// Values without a known canonical form (runtime markers, errors,
// foreign types) fall back to Value.String.
func Canon(stack []Value) string {
	parts := make([]string, len(stack))
	for i, v := range stack {
		parts[i] = CanonValue(v)
	}
	return strings.Join(parts, " ")
}

// CanonValue renders one value as canonical AQL source. See Canon.
func CanonValue(v Value) string {
	// Behavior-driven dispatch for user-defined types: if a non-
	// builtin type in v.Parent's parent chain has a non-default
	// Behavior, route through it. This is how user-installed canon
	// bodies (`behave canon/q (fn [[T] [String] [body]])`) flow
	// into eng.Canon.
	//
	// Built-in Behaviors (listFormatBehavior, mapFormatBehavior,
	// dateFormatBehavior, …) are deliberately skipped here — they
	// produce Value.String's debug form (e.g. time-domain renderings,
	// bare atoms) which doesn't match Canon's source-shape conventions
	// (e.g. `name/q` atoms, quoted strings). CanonValue's own switch
	// below preserves those.
	if v.Data != nil && v.Parent != nil {
		for t := v.Parent; t != nil; t = t.Parent {
			if t.Origin == OriginBuiltin {
				continue
			}
			if t.Behavior == nil || t.Behavior == DefaultBehavior {
				continue
			}
			if _, ok := t.Behavior.(formatDelegatesToDefault); ok {
				continue
			}
			return t.Behavior.Format(v)
		}
	}
	switch {
	case IsNone(v):
		return "none"
	case v.Data == nil:
		if t := typeNodeOf(v); t != nil {
			if name := TypeNameByID(t.ID); name != "" {
				return name
			}
			return t.Leaf()
		}
		return "none"
	case v.IsDepScalar():
		return v.String()
	case v.Parent.ConformsTo(TBigInteger):
		n, _ := AsBigInteger(v)
		return FormatBigInteger(n)
	case v.Parent.ConformsTo(TBigDecimal):
		d, _ := AsBigDecimal(v)
		return FormatBigDecimal(d)
	case v.Parent.ConformsTo(TInteger):
		n, _ := AsInteger(v)
		return strconv.FormatInt(n, 10)
	case v.Parent.ConformsTo(TFloat):
		f, _ := AsFloat(v)
		return FormatFloat(f)
	case v.Parent.ConformsTo(TString):
		s, _ := AsString(v)
		return canonString(s)
	case v.Parent.ConformsTo(TBoolean):
		b, _ := AsBoolean(v)
		if b {
			return "true"
		}
		return "false"
	case v.Parent.Equal(TAtom):
		s, _ := AsAtom(v)
		return s + "/q"
	case IsFlexList(v):
		// Round-trippable source form — a plain `[...]` would parse
		// back as an immutable List and lose the flexness.
		lst, _ := AsList(v)
		parts := make([]string, lst.Len())
		for i := 0; i < lst.Len(); i++ {
			parts[i] = CanonValue(lst.Get(i))
		}
		return "(flex [" + strings.Join(parts, " ") + "])"
	case IsFlexMap(v):
		m, err := AsMap(v)
		if err != nil || m == nil {
			return v.String()
		}
		parts := make([]string, m.Len())
		for i, k := range m.Keys() {
			val, _ := m.Get(k)
			parts[i] = k + ":" + CanonValue(val)
		}
		return "(flex {" + strings.Join(parts, " ") + "})"
	case v.Parent.ConformsTo(TList) && v.Data != nil:
		lst, _ := AsList(v)
		parts := make([]string, lst.Len())
		for i := 0; i < lst.Len(); i++ {
			parts[i] = CanonValue(lst.Get(i))
		}
		body := "[" + strings.Join(parts, " ") + "]"
		if v.Quoted {
			return "(quote " + body + ")"
		}
		return body
	case v.Parent.Equal(TMap) && v.Data != nil:
		m, err := AsMap(v)
		if err != nil || m == nil {
			return v.String()
		}
		parts := make([]string, m.Len())
		for i, k := range m.Keys() {
			val, _ := m.Get(k)
			parts[i] = k + ":" + CanonValue(val)
		}
		return "{" + strings.Join(parts, " ") + "}"
	case IsReach(v):
		return canonReach(v)
	case isFnDefValue(v):
		// A function value participates in the total order (cmp/sort),
		// so its canon form must DISCRIMINATE between distinct fns —
		// two same-shaped-but-different-body predicates must not collapse
		// to one string. Render the name plus each sig's params, returns,
		// and body. Deliberately excludes Registry and Captured: a fn's
		// closure environment is not part of its identity for ordering,
		// and dumping it would spill the module exports map (the leak
		// formatFnDef fixed for String()). See canonFnDef.
		fd, _ := v.Data.(FnDefInfo)
		return canonFnDef(fd)
	default:
		return v.String()
	}
}

// canonReachToken renders one receiver/key token of a Reach as source:
// a Word as its bare name (m.a, not m.a/q), a ParenExpr / nested Reach
// recursively, anything else via CanonValue.
func canonReachToken(v Value) string {
	switch {
	case IsWord(v):
		w, _ := AsWord(v)
		return w.Name
	case IsParenExpr(v):
		toks, _ := AsParenExpr(v)
		return "(" + canonReachTokens(toks) + ")"
	case IsReach(v):
		return canonReach(v)
	default:
		return CanonValue(v)
	}
}

// canonReachTokens renders a token sequence (receiver / computed-key / paren
// body) as source, each token via canonReachToken so words stay bare.
func canonReachTokens(toks []Value) string {
	parts := make([]string, len(toks))
	for i, t := range toks {
		parts[i] = canonReachToken(t)
	}
	return strings.Join(parts, " ")
}

// canonReach renders a Reach back to its dotted surface — m.a.b, m!.x,
// m.'k', m.(expr), (expr).k — the read∘print round-trip (design/REACH.10.md
// §6). A Quoted (codequote-captured) reach wraps in (codequote …) so it
// round-trips, mirroring the list quote convention.
func canonReach(v Value) string {
	info, err := AsReach(v)
	if err != nil {
		return v.String()
	}
	var b strings.Builder
	switch len(info.Receiver) {
	case 0:
		// receiverless reach (a lens): the reserved `$` sentinel receiver,
		// so `read ∘ print` round-trips ($.name parses back to a lens).
		b.WriteString("$")
	case 1:
		b.WriteString(canonReachToken(info.Receiver[0]))
	default:
		b.WriteString("(" + canonReachTokens(info.Receiver) + ")")
	}
	for _, seg := range info.Segments {
		if seg.Getr {
			b.WriteString("!.")
		} else {
			b.WriteString(".")
		}
		if seg.Computed {
			b.WriteString("(" + canonReachTokens(seg.KeyExpr) + ")")
		} else {
			b.WriteString(canonReachToken(seg.KeyLit))
		}
	}
	if v.Quoted {
		return "(codequote " + b.String() + ")"
	}
	return b.String()
}

// canonFnDef renders a function value's discriminating canonical form:
// its name plus the params / returns / body of each authored signature.
// Used only by CanonValue (the ordering / structural-compare surface),
// so unlike formatFnDef it must distinguish fns that String() renders
// identically. It never touches FnDefInfo.Registry or .Captured.
func canonFnDef(fd FnDefInfo) string {
	var b strings.Builder
	b.WriteString("fn ")
	b.WriteString(fd.Name)
	b.WriteByte('[')
	sigs := fd.OwnSigs()
	for i := range sigs {
		if i > 0 {
			b.WriteByte(' ')
		}
		sig := &sigs[i]
		b.WriteByte('[')
		for j, p := range sig.Params {
			if j > 0 {
				b.WriteByte(' ')
			}
			b.WriteString(p.Name)
			if p.Type != nil {
				b.WriteByte(':')
				b.WriteString(p.Type.String())
			}
		}
		b.WriteString("][")
		for j, r := range sig.Returns {
			if j > 0 {
				b.WriteByte(' ')
			}
			if r != nil {
				b.WriteString(r.String())
			}
		}
		b.WriteString("][")
		for j, t := range sig.Body {
			if j > 0 {
				b.WriteByte(' ')
			}
			b.WriteString(CanonValue(t))
		}
		b.WriteString("]")
	}
	b.WriteByte(']')
	return b.String()
}

File: eng/ts/src/canon.ts
// Canon renders values as canonical AQL source — the string a spec row's
// `expected` column is compared against. Ported from eng/go/canon.go +
// eng/go/value.go::FormatFloat so the TS engine's output matches the Go
// reference row-for-row.
//
// PARITY NOTE: this covers the value kinds the eng/spec corpus reaches
// through the current TS engine (none, type literals, scalars, atoms,
// lists, fn defs). Map / BigInteger / Reach / Flex / DepScalar branches
// are added by their owning port increments.
import { TAtom, TBoolean, TFloat, TInspect, TInteger, TList, TMap, TNone, TPath, TString } from './type.ts'
import type { FnDefInfo, XmlElement } from './value.ts'
import { ChildType, OrderedMap, Value } from './value.ts'

// canonXml renders an XML element, normalising an empty element to the
// self-closing form (<br></br> → <br/>).
function canonXml(e: XmlElement): string {
  const attrs = e.attrs.map((a) => ` ${a.name}="${a.value}"`).join('')
  if (e.children.length === 0) return `<${e.tag}${attrs}/>`
  const body = e.children.map((c) => (typeof c === 'string' ? c : canonXml(c))).join('')
  return `<${e.tag}${attrs}>${body}</${e.tag}>`
}

/**
 * canonString renders a string payload as parseable AQL source. Plain
 * content uses single quotes; content with a single quote switches to
 * double quotes; content with both quote kinds, backslashes, or control
 * characters falls back to single quotes with backslash escapes.
 * Mirrors eng/go/canon.go::canonString.
 */
export function canonString(s: string): string {
  const hasSingle = s.includes("'")
  const hasEscape = /[\\\n\t\r]/.test(s)
  if (!hasSingle && !hasEscape) return `'${s}'`
  if (!s.includes('"') && !hasEscape) return `"${s}"`
  let b = "'"
  for (const r of s) {
    switch (r) {
      case '\\':
        b += '\\\\'
        break
      case "'":
        b += "\\'"
        break
      case '\n':
        b += '\\n'
        break
      case '\t':
        b += '\\t'
        break
      case '\r':
        b += '\\r'
        break
      default:
        b += r
    }
  }
  return b + "'"
}

/**
 * formatFloat renders a float with a guaranteed decimal point so it
 * stays visually distinct from Integer, using the shortest round-trip
 * representation. Mirrors eng/go/value.go::FormatFloat.
 */
export function formatFloat(f: number): string {
  if (Number.isNaN(f)) return 'nan'
  if (f === Infinity) return 'inf'
  if (f === -Infinity) return '-inf'
  // Negative zero is a distinct bit pattern (IEEE) and renders as -0.0.
  if (f === 0 && 1 / f === -Infinity) return '-0.0'

  if (f !== 0) {
    const a = Math.abs(f)
    if (a >= 1e21 || a < 1e-10) {
      return formatExponential(f)
    }
  }
  let s = toFixedShortest(f)
  if (!/[.eE]/.test(s)) s += '.0'
  return s
}

// toFixedShortest renders a float in plain decimal ('f') form with the
// shortest precision that round-trips — the equivalent of Go's
// strconv.FormatFloat(f, 'f', -1, 64). JS's String() uses 'g'-style
// formatting (switching to exponent for large/small values), so coerce
// any exponent form back to plain decimal for the everyday range that
// FormatFloat keeps non-scientific.
function toFixedShortest(f: number): string {
  const s = String(f)
  if (!/[eE]/.test(s)) return s
  // Expand the JS exponential form to plain decimal.
  return expandExponential(s)
}

// formatExponential renders the Go strconv 'e' form: a mantissa with a
// sign-and-at-least-two-digit exponent (e.g. 1e+21 → "1e+21").
function formatExponential(f: number): string {
  let s = f.toExponential()
  // JS gives "1e+21"; Go gives "1e+21" too. Normalise the exponent to a
  // leading sign (JS already includes it) — no zero-padding needed to
  // match Go's -1 precision 'e' form for the magnitudes the specs use.
  // Drop a trailing ".0"-style mantissa? Go keeps the shortest mantissa.
  s = s.replace(/e([+-])(\d)$/, 'e$10$2')
  return s
}

// expandExponential turns a JS exponential literal into plain decimal.
function expandExponential(s: string): string {
  const neg = s.startsWith('-')
  if (neg) s = s.slice(1)
  const m = /^(\d+)(?:\.(\d+))?[eE]([+-]?\d+)$/.exec(s)
  if (!m) return (neg ? '-' : '') + s
  const intPart = m[1]!
  const fracPart = m[2] ?? ''
  const exp = Number.parseInt(m[3]!, 10)
  const digits = intPart + fracPart
  const pointPos = intPart.length + exp
  let out: string
  if (pointPos <= 0) {
    out = '0.' + '0'.repeat(-pointPos) + digits
  } else if (pointPos >= digits.length) {
    out = digits + '0'.repeat(pointPos - digits.length)
  } else {
    out = digits.slice(0, pointPos) + '.' + digits.slice(pointPos)
  }
  // Trim a trailing fractional zero run produced by the expansion.
  if (out.includes('.')) out = out.replace(/0+$/, '').replace(/\.$/, '')
  return (neg ? '-' : '') + out
}

/** Canon renders a stack of values as canonical AQL source. */
export function canon(stack: Value[]): string {
  return stack.map(canonValue).join(' ')
}

/** canonValue renders one value as canonical AQL source. */
export function canonValue(v: Value): string {
  // The `none` value renders lowercase; a bare type literal (including
  // the `None` type) renders as its user-facing leaf name.
  if (v.isNone()) return 'none'
  if (v.data === null) {
    return v.vType.leaf()
  }
  if (v.vType.matches(TInteger)) {
    return v.asInteger().toString()
  }
  if (v.vType.matches(TFloat)) {
    return formatFloat(v.asFloat())
  }
  if (v.vType.matches(TString)) {
    return canonString(v.asString())
  }
  if (v.vType.matches(TBoolean)) {
    return v.asBoolean() ? 'true' : 'false'
  }
  if (v.vType.equal(TAtom)) {
    return `${v.asAtom()}/q`
  }
  if (v.vType.equal(TPath) && v.data !== null && typeof v.data === 'object' && 'segments' in v.data) {
    const p = v.data as { segments: string[]; abs: boolean }
    return (p.abs ? '/' : '') + p.segments.join('/')
  }
  if (v.data instanceof ChildType) {
    const ct = v.data
    const open = v.isTypedMap() ? '{' : '['
    const close = v.isTypedMap() ? '}' : ']'
    const parts = [`:${canonValue(ct.child)}`, ...ct.elements.map(canonValue)]
    return `${open}${parts.join(' ')}${close}`
  }
  if (v.vType.matches(TList) && Array.isArray(v.data)) {
    const body = `[${v.asList().map(canonValue).join(' ')}]`
    return v.quoted ? `(quote ${body})` : body
  }
  if (v.vType.equal(TMap) && v.data instanceof OrderedMap) {
    const m = v.data
    const parts = m.sortedKeys().map((k) => `${k}:${canonValue(m.get(k)!)}`)
    return `{${parts.join(' ')}}`
  }
  // An inspection map renders in insertion order with bare word values
  // (e.g. `kind:native`), unlike a plain map.
  if (v.vType.equal(TInspect) && v.data instanceof OrderedMap) {
    const m = v.data
    const parts = m.keys().map((k) => {
      const val = m.get(k)!
      const rendered = val.isWord() ? val.asWord().name : canonValue(val)
      return `${k}:${rendered}`
    })
    return `{${parts.join(' ')}}`
  }
  if (v.isXml()) {
    return canonXml(v.data as XmlElement)
  }
  if (v.isFnDef()) {
    return canonFnDef(v.asFnDef())
  }
  // Disjunct / Enum: alternatives joined by '|' (atoms render bare,
  // type literals as their leaf).
  if (v.isDisjunct()) {
    return v
      .asDisjunct()
      .alternatives.map((a) =>
        a.vType.equal(TAtom) && a.data !== null
          ? a.asAtom()
          : a.data === null
            ? a.vType.leaf()
            : canonValue(a),
      )
      .join('|')
  }
  return v.toString()
}

// canonFnDef renders a function value's discriminating canonical form.
// The TS FnDefInfo carries a single params/returns/body triple (the
// spec subset), so this renders that one signature. Mirrors the shape
// of eng/go/canon.go::canonFnDef for a single authored sig.
function canonFnDef(fd: FnDefInfo): string {
  const sigs = fd.sigs
    .map((s) => {
      const params = s.params.map((p) => `${p.name}:${p.type.toString()}`).join(' ')
      const returns = s.returns.map((r) => r.toString()).join(' ')
      const body = s.body.map(canonValue).join(' ')
      return `[${params}][${returns}][${body}]`
    })
    .join(' ')
  return `fn [${sigs}]`
}
